shell/platform/darwin/macos/framework/Source/FlutterViewController.mm

```diff
@@ -60,6 +60,17 @@ void Reset() {
   }
 };
 
+/**
+ * State tracking for keyboard events, to adapt between the events coming from the system and the
+ * events that the embedding API expects.
+ */
+struct KeyboardState {
+  /**
+   * The last known pressed modifier flag keys.
+   */
+  uint64_t previously_pressed_flags = 0;
```

Comment:
Is there any way to get the current state on startup?  What happens if I start my app while holding down the shift key?

Reply:
Yeah, we can get them with [NSEvent modifierflags]. I just need to call that in the event were the view captures focus. I'll add that.

Reply:
I looked a bit more into this. I don't think that FlutterViewController can be notified when it becomes the key window. That would be the NSWindow, which is created and controlled by the runner. I think I would have to expose a OnBecomeKeyWindow callback from the ViewController, and call it from the Runner once it gets notified. 

@stuartmorgan do you know of a better way to do this?

Reply:
I believe you could just watch for `NSWindowDidBecome/ResignKey` notifications and compare the window to `self.window` when you get them. (If you instead register with the current window as the object you'll need to update your registration on any window change, which is probably more trouble than it's worth.)

Do we actually need to monitor for this though? If the goal is to prevent stuck modifier keys maybe it would be simpler to just check on every event whether there is a mismatch and synthesize events accordingly. Otherwise we'll need to revisit this when we support add-to-app, where hybrid windows can exist where being key doesn't mean the view is getting key events; we'd need to watch responder status too. And there may be other edge cases

Reply:
--  it would be simpler to just check on every event whether there is a mismatch and synthesize events accordingly
Mismatch between which values? The problem is that (AFAIK) there is no way to get the "previously pressed flags" when `flagsChanged:` is called, so we can't tell if the event is a key down, or a "one key up, but another one still down". In the case of user pressing and holding shift+command in another window, then clicking on the flutter app and releasing command, it will count it as a down event.

Reply:
I assume the goal is to prevent stuck modifiers, so the previous value would be whatever we last reported to Flutter.

The other option would be to change the framework logic to not expect every key event to come to it, since that's not how desktop OSes work. The fact that things can be stuck is, I would argue, a framework bug that we're trying to work around in the engine.

Reply:
Discussed offline with Greg. Will land this for now and filed https://github.com/flutter/flutter/issues/43309 to figure out what the best way to solve it is, potentially from the framework.